test/meta/project.json

```diff
@@ -94,6 +94,22 @@
                 "struct_w_bitmask",
                 "bitmask_w_short_notation"
             ]
+        }, {
+            "id": "RuntimeTypes",
+            "testcases": [
+                "trivial_struct",
+                "ctor",
+                "dtor",
+                "move",
+                "copy",
+                "combined",
+                "trivial_array",
+                "array_ctor",
+                "array_dtor",
+                "array_move",
+                "array_copy",
+                "mixed"
+            ]            
```

Comment:
### Tests with strings:
I believe these tests would be out of scope for RTT. `ecs_string_t` is not treated by RTT in any special way. From its point of view it is just a type that happens to have some hooks. RTT's responsibility is to make sure the underlying type's hooks are called when necessary, regardless of what type. To that effect, there is a mock type in the test cases that simulates something that holds a resource like an `ecs_string_t`. The mock allows to prove the hooks were called when they are supposed to, that's all RTT's tests care about. In the case of `ecs_string_t` whether its hooks actually do their job it is the responsibility of `ecs_string_t`'s unit tests.

### Tests with opaque types:

"root" opaque types: Prior to becoming opaques, these have `EcsType.existing == true`, defined via `ECS_COMPONENT`, and therefore get only a `flecs_default_ctor` by default (or keep whatever is set) and is not affected by this new code. There is a test that proves opaques are basically ignored.

Opaque types as RTT struct members, array and vector underlying types: Again, these are just types that happen to have hooks (or not). RTT code does not look at the `EcsType` of underlying types, it just cares about that type having lifecycle hooks or not. The mock in the tests simulate these hooks.

Therefore, I could say that a type being opaque or not does not affect lifecycle, only serialization.

### Tests of vectors of arrays of structs of nested structs with arrays of vectors of strings inside a struct with vectors 😅:
This would be a lot of fun, but the reality is that a RTT array, RTT vector or RTT struct itself becomes a type that has some hooks. Like a LEGO brick you can combine them however you like. Testing is focused on making sure the right hook is called when necessary, otherwise the list of test cases is infinite and does not really prove anything.
That said, the test case "mixed" tests an array of a struct that has arrays of structs, vectors of structs and nested structs, just for fun 🙂.

Reply:
> Like a LEGO brick you can combine them however you like. Testing is focused on making sure the right hook is called when necessary, otherwise the list of test cases is infinite and does not really prove anything.

These tests would prove that you _can_ combine things as lego blocks without something breaking. Even if it doesn't exercise different code paths today, it might after a refactor that has explicit behavior for vectors/strings/... (like what we discussed on discord).

These tests aren't for fun, they are to prevent future regressions :)

Reply:
To solve this I wrote an "universal tester". You just define the structs, however complex you like, and a query picks them up and runs the tests on them. For exampe:

```cpp
    ecs_entity_t nightmare_struct = ecs_struct(
        world,
        {.entity = test_entity(world, 0, "nightmare_structº"),
         .members = {
             {.name = "a", .type = vector_of_strings},
             {.name = "b", .type = ecs_id(ecs_i32_t)},
             {.name = "c",
              .type = ecs_struct(
                  world,
                  {.members = {
                       {.name = "a",
                        .type = ecs_struct(
                            world, {.members = {{.name = "a",
                                                 .type = array_of_strings},
                                                {.name = "b",
                                                 .type = ecs_id(ecs_i32_t)}}})},
                       {.name = "b",
                        .type = ecs_struct(
                            world,
                            {.members =
                                 {{.name = "a", .type = vector_of_ints},
                                  {.name = "b", .type = ecs_id(ecs_i32_t)},
                                  {.name = "c",
                                   .type = ecs_struct(
                                       world,
                                       {.members =
                                            {{.name = "a",
                                              .type = vector_of_ints},
                                             {.name = "b",
                                              .type = ecs_id(ecs_i32_t)},
                                             {.name = "c",
                                              .type = vector_of_ints}}})}}})},
                       {.name = "c", .type = vector_of_strings}}})}}});
```

The good thing is that you can easily add any further types you want tested.

Reply:
> Can you add tests for:
> 
> * struct with strings
> * nested struct with strings
> * struct with array of strings
> [...]

Added these as entries in the universal tester.